memgpt/agent.py

```diff
@@ -167,37 +168,81 @@ def initialize_message_sequence(
 class Agent(object):
     def __init__(
         self,
-        agent_state: AgentState,
         interface: AgentInterface,
+        # agents can be created from providing agent_state
+        agent_state: Optional[AgentState] = None,
+        # or from providing a preset (requires preset + extra fields)
+        preset: Optional[Preset] = None,
+        created_by: Optional[uuid.UUID] = None,
+        name: Optional[str] = None,
+        llm_config: Optional[LLMConfig] = None,
+        embedding_config: Optional[EmbeddingConfig] = None,
         # extras
         messages_total: Optional[int] = None,  # TODO remove?
         first_message_verify_mono: bool = True,  # TODO move to config?
     ):
+
+        # An agent can be created from a Preset object
+        if preset is not None:
+            assert agent_state is None, "Can create an agent from a Preset or AgentState (but both were provided)"
+            assert created_by is not None, "Must provide created_by field when creating an Agent from a Preset"
+            assert llm_config is not None, "Must provide llm_config field when creating an Agent from a Preset"
+            assert embedding_config is not None, "Must provide embedding_config field when creating an Agent from a Preset"
+
+            # if agent_state is also provided, override any preset values
+            init_agent_state = AgentState(
+                name=name if name else create_random_username(),
+                user_id=created_by,
+                persona=preset.persona,
+                human=preset.human,
+                llm_config=llm_config,
+                embedding_config=embedding_config,
+                preset=preset.name,  # TODO link via preset.id instead of name?
+                state={
+                    "persona": preset.persona,
+                    "human": preset.human,
+                    "system": preset.system,
+                    "functions": preset.functions_schema,
+                    "messages": None,
+                },
+            )
+
+        # An agent can also be created directly from AgentState
+        elif agent_state is not None:
+            assert preset is None, "Can create an agent from a Preset or AgentState (but both were provided)"

```

Comment:
@sarahwooders regarding:

> in case of an empty agent_state.state, read from the preset instead

If we wanted to support this, we could comment out this assert (`assert preset is None`), and instead do something like:
```python
if preset is not None and agent_state.state is None:
    agent_state.state = {
                    "persona": preset.persona,
                    "human": preset.human,
                    "system": preset.system,
                    "functions": preset.functions_schema,
                    "messages": None,
    },
```

Reply:
However, I'm not sure if this is really necessary, and I think the current setup* may be fine.

*you can either pass an `AgentState` to the init of `Agent`, or you can pass a `preset` + a few other things (e.g. `created_by`), but you can't pass both at the same time.